TheAlgorithms/Python
# improve maintainability index mi 55.888 -> 91.1844

### **Describe your change:**

This is a POC to demonstrate measuring and improving maintainability (https://github.com/TheAlgorithms/Python/issues/2128#issuecomment-703119780)

Measured with wily:
Ronnys-MacBook-Pro:Python ronny$ wily diff strings/lower.py
Using default metrics ['maintainability.mi', 'halstead.h1', 'cyclomatic.complexity', 'raw.loc']
Comparing current with b934da4 by Dhruv on 2020-10-04.
```
╒═══════════════════════════════════════╤═════════════════════════╤═══════════════════╤═════════════════════════╤═════════════════╕
│ File                                  │ Maintainability Index   │ Unique Operands   │ Cyclomatic Complexity   │ Lines of Code   │
╞═══════════════════════════════════════╪═════════════════════════╪═══════════════════╪═════════════════════════╪═════════════════╡
│ strings/lower.py                      │ 55.888 -> 91.1844       │ 3 -> 3            │ 3 -> 7                  │ 26 -> 30        │
├───────────────────────────────────────┼─────────────────────────┼───────────────────┼─────────────────────────┼─────────────────┤

```
The following changes improved the maintainability index from 55.888 to 91.1844:
- extracted method for chr(ord(capital) + 32)
- replace if operator with 2x '<=' by a const CAPITALS and 'in' operator
- rename 'word' to 'string' 
- remove redundant comments
- replace doctest comments by real test code

For manacher.py:
- fix bug: should return '' for input ''
- add some more tests
- type-hinting
- better naming
- ref to wikipedia
- maintainability 59.4401 -> 74.4849
